### ISSUE TITLE
Migration causing upgrade to fail

### ISSUE THREAD

**Issue by KDederichs** [pdsinterop/solid-nextcloud#107]
Just FYI:

The migration that's part of 0.6.0 is causing updates to fail with:
```
Exception: Database error when running migration 000000Date20220923134100 for app solid An exception occurred while executing a query: SQLSTATE[42000]: Syntax error or access violation: 1071 Specified key was too long; max key length is 3072 bytes
```

on MySql 10 databases with default config.

**Comment by JurgenG** [comment]
I can confirm this. Having the same issue on our installation.

**Comment by Potherca** [comment]
Hi @KDederichs! Thank you for reporting (and @JurgenG for confirming) this issue.

A [quick search on StackOverflow](https://stackoverflow.com/search?q=mysql+1071+is%3Aquestion) gives similar reasons (InnoDB limits) as root cause. A lot of folks usggest using setting a string limit in the migration file using `Schema::defaultStringLength(191);`, but I am not yet fully convinced this is the best solution...

I'll try to reprocude this locally and see which fix is most effective.

**Comment by Potherca** [comment]
<sup>(Oops, wrong milestone)<sup>

**Comment by JurgenG** [comment]
Is there a reason we need key lengths of more than 3072 bytes?

**Comment by Potherca** [comment]
The max length we need is 2048 (for the URL / URI field[^1]) but depending on charset collation (for instance UTF8 or UTF8-MB4) that can mean 8192 bytes are actually needed in order to store 2048 characters.

But until I can trace where this is trigger by the DB Migration code I can't make an informed decision if the best solution is to set the collation in the PHP code, or to force the default string length, or something else entirely.

[^1]: Although the RFC/Spec does not define a maximum length, for historical reasons (namely Internet Explorer and the [XML Sitemap Protocol](https://www.sitemaps.org/protocol.html)) 2048 characters has become the convention.

**Comment by poef** [comment]
https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop-11#section-11.1

"To prevent multiple uses of the same DPoP proof servers can _store, in the context of the target URI, the jti value of each DPoP proof for the time window in which the respective DPoP proof JWT would be accepted and decline HTTP requests to the same URI for which the jti value has been seen before_.  In order to guard against memory exhaustion attacks a server that is tracking jti values should reject DPoP proof JWTs with unnecessarily large jti values or store only a hash thereof."

**Comment by Potherca** [comment]
As we need both the domain and the path, what _would_ work is to hash the input and store that, rather than storing the actual URL. (for instance using SHA1)

This is something I have been thinking about anyway, for other reasons (security/privacy but also to have a fixed length db field).

@poef Any opinions regarding that approach?

**Comment by poef** [comment]
@potherca Yes, that is an excellent solution!

**Comment by Potherca** [comment]
Then I'll do some local testing and prepare a `v0.7.1` release.

**Comment by Potherca** [comment]
This issue _should_ be resolved with release `v0.7.1`. Feedback welcome!

**Comment by KDederichs** [comment]
I'll give it a try when it evenutally hits the nextcloud app store
